contracts/examples/rewards-distribution/src/rewards_distribution.rs

```diff
@@ -226,32 +237,140 @@ pub trait RewardsDistribution:
 
     #[payable("*")]
     #[endpoint(claimRewards)]
-    fn claim_rewards(&self) {
+    fn claim_rewards(
+        &self,
+        raffle_id_start: u64,
+        raffle_id_end: u64,
+        reward_tokens: MultiValueEncoded<MultiValue2<EgldOrEsdtTokenIdentifier, u64>>,
+    ) {
         let nfts = self.call_value().all_esdt_transfers();
-        let nft_token_identifier = self.nft_token_identifier().get();
-        require!(!nfts.is_empty(), "Missing payment");
-        let mut total = BigUint::zero();
-        for nft in &nfts {
-            require!(
-                nft.token_identifier == nft_token_identifier,
-                "Invalid payment"
+        self.validate_nft_payments(&nfts);
+        self.validate_raffle_id_range(raffle_id_start, raffle_id_end);
+
+        let caller = self.blockchain().get_caller();
+        for reward_token_pair in reward_tokens.into_iter() {
+            let (reward_token_id, reward_token_nonce) = reward_token_pair.into_tuple();
+            self.claim_reward_token(
+                &caller,
+                raffle_id_start,
+                raffle_id_end,
+                &reward_token_id,
+                reward_token_nonce,
+                &nfts,
             );
-            total += self.rewards(nft.token_nonce).take();
         }
-        let caller = self.blockchain().get_caller();
+        self.send().direct_multi(&caller, &nfts);
+    }
+
+    fn claim_reward_token(
+        &self,
+        caller: &ManagedAddress,
+        raffle_id_start: u64,
+        raffle_id_end: u64,
+        reward_token_id: &EgldOrEsdtTokenIdentifier,
+        reward_token_nonce: u64,
+        nfts: &ManagedVec<EsdtTokenPayment>,
+    ) {
+        let mut total = BigUint::zero();
+
+        for raffle_id in raffle_id_start..=raffle_id_end {
+            for nft in nfts {
+                if self.try_claim(raffle_id, reward_token_id, reward_token_nonce, &nft) {
+                    total += self.compute_claimable_amount(
+                        raffle_id,
+                        reward_token_id,
+                        reward_token_nonce,
+                        nft.token_nonce,
+                    )
+                }
+            }
+        }
         if total > 0 {
-            self.send().direct_egld(&caller, &total);
+            self.send()
```

Comment:
Should probably just return this payment, and accumulate them all in a ManagedVec in the caller function, to then use `direct_multi`.

And of course, separate "egld_amount` or something, since that wouldn't work otherwise.